torchao/prototype/dtypes/bitnet.py

```diff
@@ -0,0 +1,121 @@
+        return BitnetTensor(elem)
+
+    @classmethod
+    def from_unpacked(cls, unpacked: torch.Tensor) -> "BitnetTensor":
+        return cls(pack_uint2(unpacked))
+    
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
+        def allowed_subclasses(type):
+            return (
+                issubclass(cls, type) or
+                issubclass(torch._subclasses.fake_tensor.FakeTensor, type) or 
+                issubclass(torch._subclasses.functional_tensor.FunctionalTensor, type)
+            )
+        
+        if not all(allowed_subclasses(t) for t in types):
+            return NotImplemented("Bitnet, Up to the next one to handle")
+
+        if func in BITNET_OPS_TABLE:
+            return BITNET_OPS_TABLE[func](func, args, kwargs)
+        raise NotImplementedError(f"Bitnet dispatch: attempting to run {func}, this is not supported")
+    
+    @classmethod
+    def from_float(cls, w: torch.Tensor):
+        w_int2 = _quantize_int2(w, torch.uint2).to(device=w.device)
+        return w_int2
+
+@implements([torch.ops.aten.mm.default])
+def mm(func, args, kwargs):
+    x, weight = args
+    y = torch.mm(x, weight.to(torch.int8).to(x.device).to(x.dtype))
+    return y
+
+@implements([torch.ops.aten.addmm.default])
+def addmm(func, args, kwargs):
+    bias, x, weight = args
+    y = torch.addmm(bias, x, weight.to(torch.int8).to(x.device).to(x.dtype))
+    if bias is not None:
+        y += bias
+    return y
+
+@implements([torch.ops.aten.t.default])
+def t(func, args, kwargs):
+    (tensor,) = args
+    unpacked = unpack_uint2(tensor.elem).to(tensor.device)
+    transposed = unpacked.t()
+    return BitnetTensor(pack_uint2(transposed))
+
+@implements([torch.ops.aten.detach.default])
+def detach(func, args, kwargs):
+    (tensor,) = args
+    return tensor
+
+@implements([torch.ops.aten.to.dtype])
+def to_dtype(func, args, kwargs):
+    (tensor, dtype) = args
+    if dtype == torch.int8:
+        return unpack_uint2(tensor.elem).view(torch.uint8) - 1
+    elif dtype in (torch.float, torch.float16, torch.bfloat16, torch.int16, torch.int32, torch.int64):
+        return unpack_uint2(tensor.elem).to(torch.int8).to(dtype)
+    elif dtype == torch.uint8:
+        return unpack_uint2(tensor.elem).view(torch.uint8)
+    elif dtype == torch.uint2:
+        return tensor.elem
+    raise NotImplementedError(f"to {dtype} not supported")
+
+@implements([torch.ops.aten._to_copy.default])
+def _to_copy(func, args, kwargs):
+    (tensor,) = args
+    dtype = kwargs["dtype"]
+    if dtype == torch.int8:
+        return BitnetTensor(unpack_uint2(tensor).view(tensor.shape).view(torch.int8) - 1)
+    elif dtype in (torch.float, torch.float16, torch.bfloat16, torch.int16, torch.int32, torch.int64):
+        return BitnetTensor(tensor.to(torch.int8).to(dtype))
+    elif dtype == torch.uint2:
+        return BitnetTensor(tensor)
+    raise NotImplementedError(f"to {dtype} not supported")
+
+if __name__ == "__main__":
```

Comment:
I'd rather have all of this functionality be in the test file

As in make sure you  can instantiate a BItNet tensor, copy it it, transpose it, multiply and convert to from the main test

Reply:
added test cases for BitnetTensor, UIntTensor